design/src/rfcs/rfc0035_client_configuration_for_orchestrator.md

```diff
@@ -0,0 +1,270 @@
+```
+The relations between the types can be illustrated in the following diagram:
+
+<img width="1367" alt="client config at different levels" src="https://user-images.githubusercontent.com/15333866/228719737-d58d5e0a-f7ca-46d4-bb22-33c7c528ba69.png">
+
+The `aws_sdk_s3::config::Config` type on the left stores the service-level configuration, which implicitly includes/shadows the AWS-level configuration. `Config` is accessible via `Handle` from a fluent builder `ListBucketsFluentBuilder`, which holds the operation level config.
+
+After the `send` method is called on the `fluent_builder` variable and before it internally calls `aws_smithy_runtime::client::orchestrator::invoke`, the fields in `self.handle.conf` and those in `self.config_override` will be put into a type map. How this is done exactly is an implementation detail outside the scope of the RFC, but we have been working on [send_v2](https://github.com/awslabs/smithy-rs/blob/b023426d1cfd05e1fd9eef2f92a21cad58b93b86/codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/client/FluentClientGenerator.kt#L331-L347) to allow for a gradual transition.
+
+We use the `aws_smithy_runtime_api::config_bag::ConfigBag` type as the type map, and we have defined a trait called `aws_smithy_runtime_api::client::runtime_plugin::RuntimePlugin` that allows trait implementors to add key-value pairs to `ConfigBag`. The `RuntimePlugin` trait is defined as follows:
+```rust
+pub trait RuntimePlugin {
+    fn configure(&self, cfg: &mut ConfigBag) -> Result<(), BoxError>;
+}
+```
+
+With this in mind, we plan to
+- Implement the `RuntimePlugin` trait for service configuration, which will put runtime components and required function parameters specified at the service level config into `ConfigBag`
+- Implement the `RuntimePlugin` trait for service configuration builders, which will put runtime components and required function parameters specified at the operation level config into `ConfigBag`.
+
+Once these changes are made, fluent builders will be able to do something like this:
+```rust
+// pseudo-code, possibly somewhere down the line of the `send_v2` method
+
+let mut config_bag = /* create an empty config bag */;
+
+// first put into the bag things from the service level config
+self.handle.conf.configure(&mut config_bag);
+
+// then put into the bag things from the operation level config
+if let Some(config_override) = self.config_override {
+    self.config_override.configure(&mut config_bag);
+}
+```
+Note that if `ConfigLoader`, `SdkConfig`, and service configurations, including all of their builders, are implemented in terms of `ConfigBag`, implementing the `RuntimePlugin` trait may become unnecessary because the fields have already been stored in the bag. In that case, however, `ConfigBag`, which is held by `Arc<Handle>`, may need to be unshared so that it can be passed as `&mut` to `aws_smithy_runtime::client::orchestrator::invoke`.
```

Comment:
My understanding (and I could be wrong on this) is that the Smithy client goes away in the new design, so the diagram may need an update. Since the concept of an `Operation` is going away, there's no way to use a Smithy client in the new design. So the fluent client will likely just have the config as a member at that point.

Reply:
You're right about that. Now that I'm looking at the diagram, it's purely based on my prior understanding, and needs to be updated. A service client, `aws_sdk_s3::client::Client`, will no longer hold on to `aws_smithy_client::Client` in the orchestrator world. The fluent client will directly invoke `aws_smithy_runtime::client::orchestrator::invoke` within `send_v2`, rather than the old way of calling smithy client's `call` method.

Reply:
You're right about that. Now that I'm looking at the diagram, it's purely based on my prior understanding, and needs to be updated. A service client, `aws_sdk_s3::client::Client`, will no longer hold on to `aws_smithy_client::Client` in the orchestrator world. The fluent client will directly invoke `aws_smithy_runtime::client::orchestrator::invoke` within `send_v2`, rather than the old way of invoking smithy client's `call` method.